### ISSUE TITLE
`sentence_capitalization` should only be applied to actual sentences

### ISSUE THREAD

**Issue by chrisgrieser** [Automattic/harper#228]
Right now, `sentence_capitalization` is applied to all comments, regardless whether that comment is actually a sentence, a word, or commented out code. This leads to *a lot* of false positives, making `sentence_capitalization` for me pretty much unusable outside of markdown files:

![Pasted image 2024-10-16 at 17 09 13@2x](https://github.com/user-attachments/assets/ab1d38dc-2693-441e-a425-ff1aab5a5462)

**Comment by grantlemons** [comment]
I agree that it creates more false-positives that it's worth right now. I'm of the opinion it should not be enabled by default until the implementation is more mature.

@elijah-potter thoughts?

**Comment by elijah-potter** [comment]
I'm so sorry I haven't been able to get to this until now. I've been swamped. You're absolutely right that it needs work. Rather than simply disable it by default, I'd rather work with you, @chrisgrieser, to get it to a place we can proud of. 

I've made some changes that will hopefully improve it.

Are you in a position where you frequently Harper from source? If not, it's high time we ran a release, so let me know if that needs to happen for you to test the changes.

**Comment by chrisgrieser** [comment]
> Are you in a position where you frequently Harper from source? 

I do not, so yeah a release would be be nice

**Comment by hippietrail** [comment]
It should have a heuristic that decides if something is a sentence or a fragment. Perhaps also an overall style detector that doesn't flag uncapitalized sentences if the style of the entire document is fully lowercase.

**Comment by elijah-potter** [comment]
> I do not, so yeah a release would be be nice

We've had multiple releases since we last spoke. Does the issue persist?

**Comment by chrisgrieser** [comment]
The issue does not occur anymore *for the most part*, however, there are still a few edge cases. For example, a comment when there is a link inside that comment. 

![Image](https://github.com/user-attachments/assets/c56b2d4c-7b19-4a4f-b909-df912ffdd710)

**Comment by elijah-potter** [comment]
@chrisgrieser, I might not be understanding your problem then.

It sounds like your original concern was that the `sentence_capitalization` rule had a lot of false-positives in comments that weren't intended to be full sentences. Is that right?

**Comment by chrisgrieser** [comment]
>  It sounds like your original concern was that the sentence_capitalization rule had a lot of false-positives in comments that weren't intended to be full sentences. Is that right?

In my understanding, a sentence is something that ends with a `.` (or `?` or `!`). In the examples above, that is not the case, yet Harper still reports them.

```
-- the quick brown fox is faster than a turtle.
-- ^ This is a sentence, as it ends with a `.`; harper should report this.

-- some text following by a link https://github.com/Automattic/harper/issues/228
-- ^ This contains a `.` in the URL, but is still not a sentence. Harper should not report this.

-- this function does foo and bar
-- ^ This is not a sentence, but a description. Harper should not report this.
```

**Comment by hippietrail** [comment]
> -- this function does foo and bar
> -- ^ This is not a sentence, but a description. Harper should not report this.
> ```

Linguistically that is a sentence, because it has a subject, a verb, and an object. But there could be an option to not flag a sentence/chunk? if it lacks both a capital letter at the start and a period at the end. In this kind of style question marks or exclamation marks may still appear because they're more powerful than a period. And in the last couple of years young people have stopped using question marks too.

**Comment by chrisgrieser** [comment]
> Linguistically that is a sentence, because it has a subject, a verb, and an object. 

True, bad example. Here is am example of text that lacks a period and also lacks a subject, but is still being reported by harper:

![Image](https://github.com/user-attachments/assets/b3adccbf-6367-4045-9165-61e5a7010faa)

**Comment by hippietrail** [comment]
> > Linguistically that is a sentence, because it has a subject, a verb, and an object.
> 
> True, bad example. Here is an example of text that lacks a period and also lacks a subject, but is still being reported by harper:
> 
> ![Image](https://github.com/user-attachments/assets/b3adccbf-6367-4045-9165-61e5a7010faa)

I think one good approach for this is to have rules that you can enable in documentation file formats and in string literals in code file formats, but disable in comments in code file formats where a more casual style is often not just accepted but appreciated.
